8350550: Preload classes from AOT cache during VM bootstrap

iklam opened the pull request:
This PR loads the classes for the boot/platform/app loaders with `AOTLinkedClassBulkLoader::preload_classes()`. This happens at the very beginning of `vmClasses::resolve_all()`, before any Java code is executed. 

- We essentially iterate over all the classes inside the `AOTLinkedClassTable` and adds them into the system dictionary using the new method `SystemDictionary::preload_class()`.
- `SystemDictionary::preload_class(..., k)` is lightweight because it's called in a single thread after all super types of `k` have been loaded. So most of the complicated work (such as place holders, circularity detection, etc) in `SystemDictionary::resolve_or_null(..., k)` can be skipped. We also don't need to call into `ClassLoader::load_class()` as the boot/platform/app loaders are well-behaved.
- In the assembly phase, we record the mirror, package, protection domain, code source, etc, of these classes. So there's no need to programmatically create them in the production run. See `HeapShared::copy_java_mirror()` and also changes in ClassLoader.java and SecureClassLoader.java.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8350550](https://bugs.openjdk.org/browse/JDK-8350550): Preload classes from AOT cache during VM bootstrap (**Enhancement** - P3)


### Reviewers
 * [Dan Heidinga](https://openjdk.org/census#heidinga) (@DanHeidinga - no project role) Review applies to [62014d4f](https://git.openjdk.org/jdk/pull/26375/files/62014d4ff1dec5181362ca0c3eb0d89facf10283)
 * [Ashutosh Mehra](https://openjdk.org/census#asmehra) (@ashu-mehra - Committer) Review applies to [9072a1df](https://git.openjdk.org/jdk/pull/26375/files/9072a1dfb00cf51aa6a9835a7c678c68e90e617f)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/26375/head:pull/26375` \
`$ git checkout pull/26375`

Update a local copy of the PR: \
`$ git checkout pull/26375` \
`$ git pull https://git.openjdk.org/jdk.git pull/26375/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 26375`

View PR using the GUI difftool: \
`$ git pr show -t 26375`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/26375.diff">https://git.openjdk.org/jdk/pull/26375.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/26375#issuecomment-3190266812)
</details>

iklam commented on src/hotspot/share/cds/cdsHeapVerifier.cpp:
```diff
@@ -227,10 +231,16 @@ class CDSHeapVerifier::CheckStaticFields : public FieldClosure {
+            return;
+          }
+          if (field_ik == vmClasses::Boolean_klass()) {
+            // TODO: check if is TRUE or FALSE
```
Yes.

iklam commented on src/hotspot/share/cds/aotLinkedClassBulkLoader.cpp:
```diff
@@ -61,6 +64,144 @@ bool AOTLinkedClassBulkLoader::class_preloading_finished() {
+
+  // Preloading is supported only for the static archive. Classes in the dynamic archive are loaded
+  // later in load_javabase_classes() and load_non_javabase_classes().
+  preload_classes_in_table(AOTLinkedClassTable::for_static_archive()->boot1(), "boot1", Handle(), CHECK);
+  preload_classes_in_table(AOTLinkedClassTable::for_static_archive()->boot2(), "boot2", Handle(), CHECK);
```
This is still meaningful as the `boot1` classes from the dynamic CDS archive are still loaded at the end of `vmClasses::resolve_all()`, after some bytecodes are executed.

iklam commented on src/java.base/share/classes/java/net/URL.java:
```diff
@@ -1758,6 +1761,11 @@ private void setSerializedHashCode(int hc) {
+    }
+
+    @AOTRuntimeSetup
+    private static void runtimeSetup() {
```
This code basically adds an entrypoint in the `SharedSecrets` class for other JDK core lib classes to call into package-private API in this package. It doesn't do anything else.

There are several other classes where we have to do the same `SharedSecrets` set-up.

```
    @AOTRuntimeSetup
    private static void runtimeSetup() {
        SharedSecrets.setJavaNetURLAccess(
                new JavaNetURLAccess() {
                    @Override
                    public URLStreamHandler getHandler(URL u) {
                        return u.handler;
                    }
                }
        );
    }
```

DanHeidinga replied:
I'm less worried about this particular `runtimeSetup` implementation and more with what it implies.  Namely that we have URL instances - with particular URLStreamHandlers associated with them - running around in the archived heap.  If in production, a different URLStreamHandler is configured for a given URL, we'll get two different sets of validation logic for assembly time URLs vs production run URLs.

Are we able to limit the protocols that we create URLs for?  I'm reaching for some way to contain the potential issue to something smaller that we can reason about

iklam replied:
I added a check that URLs can be cache only if they use the `file` and `jtr` protocols, whose `URLStreamHandlers` cannot be overridden by the application.

iklam commented on src/java.base/share/classes/jdk/internal/loader/NativeLibraries.java:
```diff
@@ -395,9 +400,13 @@ static class LibraryPaths {
-    // All native libraries we've loaded.
-    private static final Set<String> loadedLibraryNames =
+    // Holder has the fields that need to be initialized during JVM bootstrap even if
+    // the outer is aot-initialized.
+    static class Holder {
+        // All native libraries we've loaded.
+        private static final Set<String> loadedLibraryNames =
```
The assembly phase shouldn't execute any user Java code. If that happens, it can result in many undesirable side effects. Loading user native library will be just one example.

I think we need a stronger check -- make sure that no user classes are initialized at the end of the assembly phase. This check should be sufficient because executing code in any class will result in its initialization.

Maybe I should do that in a separate RFE?

DanHeidinga replied:
> I think we need a stronger check -- make sure that no user classes are initialized at the end of the assembly phase. This check should be sufficient because executing code in any class will result in its initialization.

That would be a good check.

Do we still assembly-time initialize Enums?  I thought in earlier implementations we had initialized all enum classes but I'm having trouble finding the code now.  If we do, the new "no user class initialized" check may fail on Enums

iklam commented on src/hotspot/share/cds/aotLinkedClassBulkLoader.cpp:
```diff
@@ -61,6 +64,144 @@ bool AOTLinkedClassBulkLoader::class_preloading_finished() {
+        for (int i = 0; i < ik->local_interfaces()->length(); i++) {
+          precond(ik->local_interfaces()->at(i)->is_loaded());
+        }
+      });
```
I removed this check.

iklam commented on src/hotspot/share/cds/aotOopChecker.cpp:
```diff
@@ -0,0 +1,74 @@
+
+// Make sure we are not caching objects with assumptions that can be violated in
+// the production run.
+void AOTOopChecker::check(oop obj) {
```
The bail out and error logging needs to be done inside this function (as there might be more than one reason why `obj` is unsafe). 

This is an unrecoverable error -- the caller has already found a bad oop. We cannot just throw away this oop because there are other states that point to this oop.

DanHeidinga replied:
The existing error handling - to log and bail out - are 100% right.  We're not throwing an exception so the caller still needs to unwind itself as well and a `return false;` after the `MetaspaceShared::unrecoverable_writing_error();` makes the caller more obvious

iklam replied:
`MetaspaceShared::unrecoverable_writing_error()` will terminate the VM. So we will never return to the caller.

DanHeidinga replied:
That makes sense.  Thanks for confirming.

DanHeidinga reviewed (approved):
My concerns have been addressed.  Thanks Ioi

ashu-mehra commented on src/hotspot/share/cds/aotOopChecker.cpp:
```diff
@@ -0,0 +1,74 @@
+void AOTOopChecker::check(oop obj) {
+  precond(vmClasses::URL_klass()->is_final());
+
+  if (obj->klass()->is_subclass_of(vmClasses::URL_klass())) {
```
URL is a final class, so shouldn't `obj->klass() == vmClasses::URL_klass()` be sufficient?

ashu-mehra commented on src/hotspot/share/classfile/javaClasses.cpp:
```diff
@@ -1052,9 +1061,11 @@ void java_lang_Class::set_mirror_module_field(JavaThread* current, Klass* k, Han
 
 // Statically allocate fixup lists because they always get created.
 void java_lang_Class::allocate_fixup_lists() {
```
Do we need `fixup_module_field_list` when `is_using_preloaded_classes` is true?

iklam replied:
I've added comments why fixup_module_field_list is not needed when is_using_preloaded_classes is true?

ashu-mehra commented on src/hotspot/share/memory/iterator.inline.hpp:
```diff
@@ -51,12 +50,7 @@ inline void ClaimMetadataVisitingOopIterateClosure::do_cld(ClassLoaderData* cld)
-  if (cld != nullptr) {
-    ClaimMetadataVisitingOopIterateClosure::do_cld(cld);
-  } else {
-    assert(AOTLinkedClassBulkLoader::is_pending_aot_linked_class(k), "sanity");
```
Why is the null check and assert not valid anymore?

iklam replied:
The AOT cache can have heap objects whose class are loaded by the platform/app loaders. These objects usually come from aot-linked Lambda expressions.

Before this PR, the platform/app classes are loaded after executing quite a bit of Java code (i.e.,module graph initialization). It's possible for a GC to happen and discover those heap objects, whose classes are not yet loaded (i.e., having a null `cld`).

After this PR, all boot/platform/app classes are loaded before GC can happen, so we will never run into such null `cld` anymore.

ashu-mehra replied:
That makes `is_pending_aot_linked_class` dead code. Can it be removed?

iklam commented on src/hotspot/share/cds/filemap.cpp:
```diff
@@ -1858,6 +1858,11 @@ bool FileMapInfo::validate_aot_class_linking() {
   if (header()->has_aot_linked_classes()) {
     const char* archive_type = CDSConfig::type_of_archive_being_loaded();
     CDSConfig::set_has_aot_linked_classes(true);
+    if (is_static()) {
```
There is actually a tricky situation -- static archive doesn't have `-XX:+AOTClassLinking` but the dynamic archive was created with `-XX:+AOTClassLinking`.

I have created [JDK-8367366](https://bugs.openjdk.org/browse/JDK-8367366) to remove support for  `-XX:+AOTClassLinking` for the dynamic archive. If that gets reviewed and integrated, then I will update this PR to remove `is_using_preloaded_classes()`

iklam commented:
@ashu-mehra , I've remove all the code that was needed to handle aot-linked classes from the dynamic archive but is no longer needed since [JDK-8367366](https://bugs.openjdk.org/browse/JDK-8367366). I am able to remove about 200 lines of code.

I also renamed a few functions in `AOTLinkedClassBulkLoader` to reflect the up-to-date meaning. 

`CDSConfig::is_using_preloaded_classes()` has been removed and replaced with `CDSConfig::is_using_aot_linked_classes()`

The handling of `fixup_mirror_list()` and `fixup_module_field_list()` in javaClasses.cpp have been made uniform.

Please see https://github.com/openjdk/jdk/pull/26375/commits/896ce8b9636a5534f1db487afbe27e875022fbfa

ashu-mehra commented:
@iklam thank you for updating the changes. It is much easier to follow the code now. I have just one additional comment, otherwise looks good.

ashu-mehra reviewed (approved):
lgtm
